File: src/main/scala/com/exasol/cloudetl/kafka/KafkaConsumerFactory.scala
package com.exasol.cloudetl.kafka

import com.exasol.errorreporting.ExaError

import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.KafkaException
import org.apache.kafka.common.serialization.Deserializer

/**
 * A factory class that creates Kafka consumer clients.
 */
object KafkaConsumerFactory {

  /**
   * Creates a [[org.apache.kafka.clients.consumer.KafkaConsumer]] from
   * [[KafkaConsumerProperties]] properties.
   *
   * At the moment Avro based specific {@code KafkaConsumer[String,
   * GenericRecord]} consumer is returned. Therefore, in order to define
   * the schema of [[org.apache.avro.generic.GenericRecord]] the {@code
   * SCHEMA_REGISTRY_URL} value should be provided.
   */
  def apply[K, V](
    properties: KafkaConsumerProperties,
    keyDeserializer: Deserializer[K],
    valueDeserializer: Deserializer[V]
  ): KafkaConsumer[K, V] = {
    val topic = properties.getTopic()
    try {
      new KafkaConsumer(properties.getProperties(), keyDeserializer, valueDeserializer)
    } catch {
      case exception: KafkaException =>
        throw new KafkaConnectorException(
          ExaError
            .messageBuilder("F-KCE-1")
            .message("Could not create a Kafka consumer for topic {{TOPIC}}.", topic)
            .ticketMitigation()
            .toString(),
          exception
        )
    }
  }

}

File: src/main/java/com/exasol/cloudetl/kafka/KafkaConsumerFactory.java
package com.exasol.cloudetl.kafka;

import com.exasol.errorreporting.ExaError;

import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.KafkaException;
import org.apache.kafka.common.serialization.Deserializer;

public final class KafkaConsumerFactory {
    private KafkaConsumerFactory() {
    }

    public static <K, V> KafkaConsumer<K, V> apply(final KafkaConsumerProperties properties,
            final Deserializer<K> keyDeserializer, final Deserializer<V> valueDeserializer) {
        final String topic = properties.getTopic();
        try {
            return new KafkaConsumer<>(properties.getProperties(), keyDeserializer, valueDeserializer);
        } catch (final KafkaException exception) {
            throw new KafkaConnectorException(ExaError.messageBuilder("F-KCE-1")
                    .message("Could not create a Kafka consumer for topic {{TOPIC}}.", topic)
                    .ticketMitigation()
                    .toString(), exception);
        }
    }
}
